aten/src/ATen/native/xnnpack/MaxPooling.cpp

```diff
@@ -0,0 +1,212 @@
+namespace native {
+namespace xnnpack {
+
+// Supports NHWC and NCHW FP32 max pooling with any
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+
+bool use_max_pool2d(
+    const Tensor& input,
+    const IntArrayRef kernel_,
+    const IntArrayRef padding_,
+    IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  // Make sure we are not dealing with an unorthodox configuration.
+  if (kernel_.empty() || padding_.empty() || dilation_.empty()) {
+    return false;
+  }
+
+  // Stride can be legitimately empty, in which case it is to be defaulted to kernel size.
+  if (stride_.empty()) {
+    stride_ = kernel_;
+  }
+
+  // Normalize the parameters.
+  const internal::pooling::Parameters parameters{
+    kernel_,
+    padding_,
+    stride_,
+    dilation_,
+  };
+
+  // Here are the list of conditions required for this code path to be taken:
+  // * Input must be 4D CPU float tensor with no gradients.
+  // * Kernel must be a 2D IntArrayRef containing two positive numbers.
+  //   Furthermore, 1x1 kernels are neither valid nor meaningful.
+  // * Padding must be a 2D IntArrayRef containing two non-negative numbers.
+  // * Stride must be a 2D IntArrayRef containing two positive numbers.
+  // * Dilation must be a 2D IntArrayRef containing two positive numbers.
+  // * Ceil mode is not supported and must be disabled.
+  // * output_max must be greater than output_min.
+  //   Namely, setting both output_min and output_max to 0 is not valid usage.
+  // * Finally, application of this operator to the input tensor with the given
+  //   max pool 2d parameters must result in an output tensor with a valid shape.
+
+  return xnnpack::internal::available() &&
+      // Input
+      (4 == input.dim()) &&
+      (c10::DeviceType::CPU == input.device().type()) &&
+      (kFloat == input.scalar_type()) &&
+      !input.requires_grad() &&
+      // Kernel
+      (2 == parameters.kernel.size()) &&
+      (parameters.kernel[Layout::Parameter::height] > 0) &&
+      (parameters.kernel[Layout::Parameter::width] > 0) &&
+      ((parameters.kernel[Layout::Parameter::height] *
+        parameters.kernel[Layout::Parameter::width]) > 1) &&
+      // Padding
+      (2 == parameters.padding.size()) &&
+      (parameters.padding[Layout::Parameter::height] >= 0) &&
+      (parameters.padding[Layout::Parameter::width] >= 0) &&
+      // Stride
+      (2 == parameters.stride.size()) &&
+      (parameters.stride[Layout::Parameter::height] > 0) &&
+      (parameters.stride[Layout::Parameter::width] > 0) &&
+      // Dilation
+      (2 == parameters.dilation.size()) &&
+      (parameters.dilation[Layout::Parameter::height] > 0) &&
+      (parameters.dilation[Layout::Parameter::width] > 0) &&
+      // Ceil Mode
+      !ceil_mode &&
+      // Output Min / Max
+      (output_max > output_min) &&
+      // Output
+      (pooling_output_shape(
```

Comment:
Does this function assume parameters such as kernel, padding, stride both must be symmetrical, meaning both height and width of these parameters be the same? Because we are not enforcing that anywhere else.